### PR TITLE
[FEATURE] Adds definePlugin functionality

### PR DESCRIPTION
**Pull request by etgrieco**
Proposed alternative to https://github.com/ExLibrisGroup/primo-explore-devenv/pull/89

EDIT (1/11/2019): (More verbose description now that https://github.com/ExLibrisGroup/primo-explore-devenv/pull/89 is closed).
---

I propose we implement [babel's define plugin](https://www.npmjs.com/package/babel-plugin-transform-define) in order to compile custom JavaScript based on a specified environment. We can use a standardized variable, i.e. `NODE_ENV` to specify the environment. Alternatively, developers can specify the environment with an `--environment` flag.

This plugin causes the following effects:

`gulp run --browserify --environment=development`

*Source Code*

```js
if (process.env.NODE_ENV === "development") {
  console.log('This is a development version of the JavaScript code!');
  // do something else useful here, like specifying development/staging URLs of affiliate services.
}
```

*Output Code*

```js
if (true) {
  console.log(true);
}
```

The default value for `NODE_ENV`, if never specified at build or via an `--environment` flag, will be `production`. Environments NOT specified as production will also skip the uglification process.

I.e.
```js
// buildByConcatenation...
  .pipe(process.env.NODE_ENV === 'production' ? uglify() : gutil.noop())
```

### CODE REVIEW

**Comment by etgrieco** [comment]
@meinbrodt Let me know what you think; incorporated your suggestion into https://github.com/ExLibrisGroup/primo-explore-devenv/pull/96/commits/d6d9693e2e3b4c1b5984b453c7805ec55b51ade2

Also: https://github.com/ExLibrisGroup/primo-explore-devenv/pull/96/commits/d6d9693e2e3b4c1b5984b453c7805ec55b51ade2 adds so that ONLY `production` environment is `uglified`, versus if `--environment=development`, or anything else, uglification is skipped. This makes sense for development contexts where uglification can sometimes make things a bit more difficult (even with sourcemaps) and also adds unnecessary overhead to the `gulp-watch` process.

**Comment by etgrieco** [comment]
@meinbrodt Thank you! It was a great idea.

**Comment by etgrieco** [comment]
Since NODE_ENV will produce different code ouput, https://github.com/ExLibrisGroup/primo-explore-devenv/pull/96/commits/7f37de5ca3902c7bd93f005ab97fe4d3b114e1e6 will change the naming scheme of `.zip` packages.

e.g. `NYU.zip` -> `NYU.201901171405.development.zip`

**Comment by meinbrodt** [comment]
Hi, I was wondering if any of the repository-maintainers already found some time to take a first look at this request.

File: gulp/tasks/createPackage.js [hunk 1]
```diff
@@ -55,13 +55,14 @@ gulp.task('select-view', (cb) => {
...
+    const env = process.env.NODE_ENV;
     const code = config.view();
     console.log('Creating package for : ('+code+'.zip)');
     console.log(code);
     console.log(' in  : /packages');
     console.log('\r\n');
     console.log('............................................................................................................................................');
     return gulp.src(['./primo-explore/custom/'+code,'./primo-explore/custom/'+code+'/html/**','./primo-explore/custom/'+code+'/img/**','./primo-explore/custom/'+code+'/css/custom1.css','./primo-explore/custom/'+code+'/js/custom.js'], {base: './primo-explore/custom'})
-        .pipe(zip(code+'.zip'))
+        .pipe(zip(`${code}.${new Date().toISOString().replace(/[^0-9]/g,'').slice(0, 12)}.${env === 'production' || !env ? 'production' : env}.zip`))
```
Comment (atla5) [inline gulp/tasks/createPackage.js]: this is a very welcome change. 👍

Comment (etgrieco) [reply gulp/tasks/createPackage.js]: @atla5 thanks! It works great with our CI/CD. much less confusing

**Pull request merged**